libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -3715,3 +3751,397 @@ instance HasSpec fn a => Pretty (Spec fn a) where
+  (_, TrueSpec) -> TrueSpec
+  (_, SuspendedSpec _ _) -> TrueSpec
+  (SuspendedSpec _ _, _) -> TrueSpec
+  (TypeSpec x bad1, TypeSpec y bad2) -> TypeSpec (ft x y) [f b1 b2 | b1 <- bad1, b2 <- bad2]
+  (MemberSpec [], _) -> ErrorSpec ["Null MemberSpec"]
+  (_, MemberSpec []) -> ErrorSpec ["Null MemberSpec"]
+  (MemberSpec xs, MemberSpec ys) -> MemberSpec [f x y | x <- xs, y <- ys]
+  -- This block is all (MemberSpec{}, TypeSpec{}) with MemberSpec on the left
+  (MemberSpec xs, TypeSpec (NumSpecInterval (Just i) (Just j)) bad) ->
+    MemberSpec [f x y | x <- xs, y <- [i .. j], not (elem y bad)]
+  -- Somewhat loose spec here, but more accurate then TrueSpec, it is exact if 'xs' has one element (i.e. 'xs' = [i])
+  (MemberSpec xs, TypeSpec (NumSpecInterval lo hi) bads) ->
+    -- We use the specialized version of 'TypeSpec' 'typeSpecOpt'
+    typeSpecOpt
+      (NumSpecInterval (f (minimum xs) <$> lo) (f (maximum xs) <$> hi))
+      [f x b | x <- xs, b <- bads]
+  -- we flip the arguments, so we need to flip the functions as well
+  (x, y) -> operateSpec (\a b -> f b a) (\u v -> ft v u) y x
+
+-- =================================
+-- Cardinality
+
+-- | Put some (admittedly loose bounds) on the number of solutions that
+--   'genFromTypeSpec' might return. For lots of types, there is no way to be very accurate.
+--   Here we lift the HasSpec methods 'cardinalTrueSpec' and 'cardinalTypeSpec'
+--   from (TypeSpec fn Int) to (Spec fn Int)
+cardinality :: forall fn a. (Eq a, BaseUniverse fn, HasSpec fn a) => Spec fn a -> Spec fn Integer
+cardinality TrueSpec = cardinalTrueSpec @fn @a
+cardinality (MemberSpec es) = exactSizeSpec (listSize (nub es))
+cardinality ErrorSpec {} = equalSpec 0
+-- This is a concession, there is not really a good way to know
+cardinality (TypeSpec s cant) =
+  subSpecInt (cardinalTypeSpec @fn @a s) (exactSizeSpec (listSize (filter (\c -> conformsTo @fn @a c s) cant)))
+cardinality SuspendedSpec {} = TrueSpec
+
+-- The cardinalSize of (NumSpec n) for a Bounded type 'n', can be made accurate
+-- by using the Bounds to capture exact information.
+cardinalSizeBoundedSpec :: forall n fn. (Bounded n, Integral n) => NumSpec n -> Spec fn Integer
+cardinalSizeBoundedSpec x@(NumSpecInterval _ _) = MemberSpec [n]
+  where
+    n = countSpec x
+
+-- The cardinalSize of (NumSpec Integer) is concrete, only if both the
+-- the lower and upper bounds are concrete. We can't play the game we used in cardinalSizeBoundedSpec
+-- Because Integer does not have minBound and maxBound instances.
+-- So if either NumSpec bound is Nothing, then we return TrueSpec.
+-- Just a little bit of infinity, makes the whole thing infinity.
+cardinalSizeIntegerSpec :: forall n fn. Integral n => NumSpec n -> Spec fn Integer
+cardinalSizeIntegerSpec (NumSpecInterval (Just lo) (Just hi)) =
+  if hi >= lo
+    then MemberSpec [fromIntegral @n @Integer (hi - lo)]
+    else MemberSpec [0]
+cardinalSizeIntegerSpec (NumSpecInterval _ _) = TrueSpec
```

Comment:
So there is a design decision here. All the cardinal stuff depends on the Num instances, which we debated above as being a bit dangerous because of overflow.

Here we are more accurate on Bounded types, but they are more dangerous elsewhere.

I guess we have to make a decision if we only want this type at Integer. If so all the other stuff goes away.

Reply:
Check into MaybeBounded

Reply:
Now there is one one function.
```
cardinalSizeSpec :: forall n fn. (Integral n, Num n,MaybeBounded n) => NumSpec n -> Spec fn Integer
```